src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/GraphNodes/DependenciesGraphProvider.cs

```diff
@@ -342,11 +332,39 @@ private static GraphNodeId GetTopLevelGraphNodeId(string projectPath, string mod
                 GraphNodeId.GetPartial(CodeGraphNodeIdName.File, filePath));
         }
 
-        private static readonly ConcurrentDictionary<(int id, Guid guid), string> s_iconNameCache = new ConcurrentDictionary<(int id, Guid guid), string>();
-
-        private static string GetIconStringName(ImageMoniker icon)
+        private sealed class GraphIconCache
         {
-            return s_iconNameCache.GetOrAdd((icon.Id, icon.Guid), i => $"{i.guid:D};{i.id}");
+            /// <summary>The set of icons that have already been registered.</summary>
+            private ImmutableHashSet<ImageMoniker> _registeredIcons = ImmutableHashSet<ImageMoniker>.Empty;
+
+            /// <summary>A cache of icon names, the reuse of which helps reduce allocations.</summary>
+            private readonly ConcurrentDictionary<(int id, Guid guid), string> _iconNameCache = new ConcurrentDictionary<(int id, Guid guid), string>();
```

Comment:
How many threads are going to access this cache at once?

Reply:
I'm not certain. This is the technique in use beforehand. Once populated, this is likely to be fairly read-only, so maybe an immutable interlocked pattern would be more suitable.

Reply:
I guess the correct answer is "we should measure" :), but I don't know if we want to pay for the cost of `ConcurrentDictionary` locking and memory complexity if there are not going to be a lot of parallel read/writes

Reply:
To measure meaningfully we'd need to know the level of concurrency, and whether we're seeing regular writes. I suspect writes to be up front then infrequent, after which `ImmutableInterlocked` is a great fit both in terms of memory and locking.